docs/guide.md

```diff
@@ -0,0 +1,69 @@
+
+# Guide to the Dafny Core Library
+
+_THIS DOCUMENT IS A DRAFT UNDER ACTIVE DEBATE AND EDITING_
+
+A programming language library encapsulates a useful set of basic
+functionality that programmers can then incorporate in their custom
+program without having to rewrite, debug, or verify.
+
+Like other languages, the Dafny programming language and tool set includes
+libraries that provide functionality in various domains. This document describes 
+the _Dafny Core Library_. Dafny programs are copmponed of modules; to ensure
+verification soundness, Dafny imposes a strict non-circular dependency relationship
+among modules. This Core library contains functionality that may be needed
+by other libraries, but it itself is standalone.
+
+All the supported, delivered libraries will
+- be stable --- a minimum of backwards incompatible changes
+- be verified --- proved to meet specifications and tested with expected use cases,
+with each release of the library and of the Dafny toolset
+- be efficient --- efficient in both proofs and executables that use the libraries
+
+The Dafny code for the library is written to be understandable with a minimum of inline documentation
+and is open source and available [on github](https://github.com/dafny-lang/libraries). This document is
+meant to be tutorial and example-based.
+
+## Content of the Core Library
+
+The sections below describe how to use each of the principal modules of the library:
+- [Dafny.Boxes](#sec-boxes) -- containers that enable encapsulaton of success and failure results in a single datatype
```

Comment:
Not a bad alternate name. I wonder if we could add `AtomicBox` here too (although that would mean adding a target language dependency): https://github.com/dafny-lang/dafny/blob/master/Source/DafnyRuntime/DafnyRuntimeDafny/src/dafnyRuntime.dfy#L359

Reply:
Thinking more about this, I don't think "encapsulation of success and failure results in a single datatype" describes Option very well. It can be used like that as a result type, but it's usually better to provide more information about a "failure". It's used far more commonly for optional input, and hence the name.

My personal inclination is that "Wrappers" and "Boxes" are about equally "decent" rather than "amazing" so I'm not convinced the renaming cost is worth it yet. :)

Reply:
I agree that `Boxes` is, if any better, not enough better that I really want to refactor everything.
I don't think that a new user coming to Dafny is really helped here.

If you wanted a name that really helped I would suggest `Results` or `FailureCompatibleTypes` or `ElephantTypes`.

Reply:
Is the Result datatype the most commonly used? If so there is some advantage to naming the module Result.